lib/artsy-eventservice/artsy/event_service/rabbitmq_connection.rb

```diff
@@ -3,44 +3,57 @@
 
 module Artsy
   module EventService
-    module RabbitMQConnection
-      # get a new RabbitMQ Client
-      def create_conn
-        Bunny.new(rabbitmq_url, **bunny_params)
-      end
+    class RabbitMQConnection
+      @mutex = Mutex.new
 
-      # Connect, do something and close the connection
-      def connect_to_rabbit
-        conn = create_conn
+      # Build a new connection to RabbitMQ
+      def self.build_connection
+        conn = Bunny.new(self.rabbitmq_url, **self.bunny_params)
         conn.start
-        yield(conn)
-      ensure
-        conn.stop
+        conn
+      end
+
+      # Access the connection - rebuild if closed
+      def self.get_connection
+        @connection = nil if @connection && @connection.closed?
+        @connection ||= self.build_connection
+      end
+
+      # Synchronized access to the connection and to the created channel
+      def self.get_channel
+        @mutex.synchronize do
+          begin
+            channel = self.get_connection.create_channel
+            yield channel if block_given?
+          ensure
+            channel.close if channel && channel.open?
+          end
+        end
```

Comment:
I don't see why we should synchronize channel creation if it's a resource designed to _not_ be shared between threads. That seems like a design smell. 

I don't believe we ever need to set the connection to `nil` as it automatically attempts to recover from network failure, as I mention [here](https://github.com/artsy/artsy-eventservice/pull/12#discussion_r144345704). 

All we need is an initialization hook for apps to do connection creation - eg we typically use initializers in Gravity which run prior to Puma threads being spawned. So you can keep connection creation within the flow of channel creation for single threaded apps, but additionally add an initializer method which an app like Gravity can call during its init phase. 

In that approach we shift responsibility to the client application, while this service doesn't make any assumptions about the app environment it's being loaded into, it only determines that some resources are shared within a process (connection) and some aren't (channels).

Reply:
Agreed - just reverted that commit and replaced with https://github.com/artsy/artsy-eventservice/pull/12/commits/ef389e9b5a94729fba74e307d886ec0e21e8a62c so we don't ever reset the connection to nil after initialized.  So here we get channels multiplexing on a single connection, yet it will still work as expected in a single process.

As far as initialization hooks go... it seems like the design of this library avoided initialization hooks that except when class loading.  But I am not adverse to requiring explicit initialization.  @ashkan18 your thoughts on this?

Reply:
I see, looks good. Since @ashkan18 is out we could merge it in as-is if you think there's no longer a risk of a race condition.